File: ios-app/HealthDecoder/UI/FamilyUi.swift
import SwiftUI

/// Port of `ui/FamilyUi.kt` — family member management (add/edit/remove) plus the Scan screen's
/// patient picker.

private let avatarChoices = ["👤", "🧑", "👩", "👨", "👵", "👴", "🧒", "👧", "👦", "🧓"]
private let sexChoices = ["Male", "Female", "Other"]

/// Whole years from a YYYY-MM-DD date of birth, or nil if unparseable/blank/future.
func familyAge(_ dob: String) -> Int? {
    let parts = dob.trimmingCharacters(in: .whitespaces).split(separator: "-")
    guard parts.count == 3,
          let y = Int(parts[0]), let m = Int(parts[1]), let d = Int(parts[2]) else { return nil }
    var components = DateComponents()
    components.year = y; components.month = m; components.day = d
    guard let birth = Calendar.current.date(from: components), birth <= Date() else { return nil }
    guard let age = Calendar.current.dateComponents([.year], from: birth, to: Date()).year else { return nil }
    return (0...130).contains(age) ? age : nil
}

/// One-line "Father • M • 58y" style summary for a profile.
func familySubtitle(_ p: FamilyProfile) -> String {
    var bits: [String] = []
    if !p.relation.isEmpty { bits.append(p.relation) }
    if let first = p.sex.first { bits.append(String(first).uppercased()) }
    if let age = familyAge(p.dateOfBirth) { bits.append("\(age)y") }
    return bits.joined(separator: " • ")
}

/// Lists family members with add/edit/remove. Removing is only offered for people with no
/// records (rename/merge is the tool for a mis-scanned duplicate).
struct FamilyManagerSheet: View {
    let onChanged: () -> Void

    @Environment(\.dismiss) private var dismiss
    @State private var members: [FamilyProfile] = []
    @State private var editing: FamilyProfile?
    @State private var adding = false
    @State private var blockedDeleteName: String?

    var body: some View {
        NavigationStack {
            List {
                ForEach(members) { member in
                    HStack(spacing: 10) {
                        Text(member.avatarEmoji).font(.system(size: 22))
                        VStack(alignment: .leading) {
                            Text(member.name).fontWeight(.bold)
                            let subtitle = familySubtitle(member)
                            if !subtitle.isEmpty {
                                Text(subtitle).font(.caption).foregroundColor(.secondary)
                            }
                        }
                        Spacer()
                        Button { editing = member } label: { Image(systemName: "pencil") }
                            .buttonStyle(.borderless)
                        Button(role: .destructive) { attemptDelete(member) } label: {
                            Image(systemName: "trash")
                        }
                        .buttonStyle(.borderless)
                    }
                }
                if members.isEmpty {
                    Text(tr("No family members yet. Add one to scope records per person."))
                        .font(.footnote)
                        .foregroundColor(.secondary)
                }
            }
            .navigationTitle(tr("Family / Patients"))
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .navigationBarLeading) {
                    Button(tr("DONE")) { dismiss() }
                }
                ToolbarItem(placement: .navigationBarTrailing) {
                    Button { adding = true } label: { Label(tr("Add"), systemImage: "plus") }
                }
            }
            .onAppear(perform: reload)
            .sheet(isPresented: $adding) {
                FamilyEditSheet(existing: nil) { reload(); onChanged() }
            }
            .sheet(item: $editing) { member in
                FamilyEditSheet(existing: member) { reload(); onChanged() }
            }
            .alert("Can't remove this person", isPresented: Binding(
                get: { blockedDeleteName != nil },
                set: { if !$0 { blockedDeleteName = nil } }
            )) {
                Button(tr("OK"), role: .cancel) {}
            } message: {
                Text("\(blockedDeleteName ?? "This person") still has saved reports. Rename them instead, or delete their reports first.")
            }
        }
    }

    private func reload() { members = AppSettings.familyProfiles }

    private func attemptDelete(_ member: FamilyProfile) {
        if LocalRepository.reportCount(forPatient: member.name) > 0 {
            blockedDeleteName = member.name
        } else {
            AppSettings.familyProfiles.removeAll { $0.id == member.id }
            reload()
            onChanged()
        }
    }
}

/// Add or edit one family member. `existing == nil` = add mode.
struct FamilyEditSheet: View {
    let existing: FamilyProfile?
    let onSaved: () -> Void

    @Environment(\.dismiss) private var dismiss
    @State private var name: String
    @State private var relation: String
    @State private var sex: String
    @State private var dob: String
    @State private var emoji: String
    @State private var errorMessage: String?

    init(existing: FamilyProfile?, onSaved: @escaping () -> Void) {
        self.existing = existing
        self.onSaved = onSaved
        _name = State(initialValue: existing?.name ?? "")
        _relation = State(initialValue: existing?.relation ?? "")
        _sex = State(initialValue: existing?.sex ?? "")
        _dob = State(initialValue: existing?.dateOfBirth ?? "")
        _emoji = State(initialValue: existing?.avatarEmoji ?? "👤")
    }

    var body: some View {
        NavigationStack {
            Form {
                Section(tr("Avatar")) {
                    ScrollView(.horizontal, showsIndicators: false) {
                        HStack(spacing: 6) {
                            ForEach(avatarChoices, id: \.self) { choice in
                                Text(choice)
                                    .font(.system(size: 20))
                                    .frame(width: 40, height: 40)
                                    .background(emoji == choice ? Color.medicalSurfaceVariant : Color(.secondarySystemBackground))
                                    .overlay(
                                        RoundedRectangle(cornerRadius: 10)
                                            .stroke(Color.medicalTeal, lineWidth: emoji == choice ? 2 : 0)
                                    )
                                    .clipShape(RoundedRectangle(cornerRadius: 10))
                                    .onTapGesture { emoji = choice }
                            }
                        }
                    }
                }
                Section {
                    TextField(tr("Name"), text: $name)
                    TextField(tr("Relation (e.g. Father, Self)"), text: $relation)
                }
                Section(tr("Gender")) {
                    Picker(tr("Gender"), selection: $sex) {
                        Text("—").tag("")
                        ForEach(sexChoices, id: \.self) { Text($0).tag($0) }
                    }
                    .pickerStyle(.segmented)
                }
                Section {
                    TextField(tr("Birthdate (YYYY-MM-DD)"), text: $dob)
                        .keyboardType(.numbersAndPunctuation)
                    if let age = familyAge(dob) {
                        Text("Age: \(age)").font(.caption).foregroundColor(.secondary)
                    }
                }
                if let errorMessage {
                    Text(errorMessage).foregroundColor(.red).font(.footnote)
                }
            }
            .navigationTitle(existing == nil ? "Add family member" : "Edit \(existing!.name)")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .cancellationAction) { Button(tr("Cancel")) { dismiss() } }
                ToolbarItem(placement: .confirmationAction) {
                    Button(existing == nil ? "Add" : "Save") { save() }
                        .disabled(name.trimmingCharacters(in: .whitespaces).isEmpty)
                }
            }
        }
    }

    private func save() {
        let trimmedName = name.trimmingCharacters(in: .whitespaces)
        var profiles = AppSettings.familyProfiles

        // Name is the join key to reports, so it must stay unique across members.
        let clashes = profiles.contains {
            $0.id != existing?.id && $0.name.caseInsensitiveCompare(trimmedName) == .orderedSame
        }
        if clashes {
            errorMessage = "A member with that name already exists."
            return
        }

        if let existing {
            if let index = profiles.firstIndex(where: { $0.id == existing.id }) {
                // Renaming cascades to that person's stored records, mirroring Android's
                // LocalRepository.updateFamilyMember.
                if existing.name != trimmedName {
                    LocalRepository.renamePatient(from: existing.name, to: trimmedName)
                }
                profiles[index] = FamilyProfile(
                    id: existing.id, name: trimmedName, relation: relation, avatarEmoji: emoji,
                    sex: sex, dateOfBirth: dob.trimmingCharacters(in: .whitespaces)
                )
            }
        } else {
            profiles.append(FamilyProfile(
                id: UUID().uuidString, name: trimmedName, relation: relation, avatarEmoji: emoji,
                sex: sex, dateOfBirth: dob.trimmingCharacters(in: .whitespaces)
            ))
        }
        AppSettings.familyProfiles = profiles
        onSaved()
        dismiss()
    }
}

/// Patient selector for the Scan screen: pick an existing family member, auto-detect, or a new
/// person. `value` is the chosen patient name; empty = auto-detect.
struct ScanPatientPicker: View {
    @Binding var value: String
    @State private var members: [FamilyProfile] = []
    @State private var manualEntry = false

    var body: some View {
        VStack(alignment: .leading, spacing: 6) {
            Menu {
                Button(tr("Auto-detect from report")) { manualEntry = false; value = "" }
                ForEach(members) { member in
                    Button {
                        manualEntry = false
                        value = member.name
                    } label: {
                        Text("\(member.avatarEmoji)  \(member.name)")
                    }
                }
                Button(tr("➕  New person…")) { manualEntry = true; value = "" }
            } label: {
                HStack {
                    Image(systemName: "person")
                    Text(manualEntry ? "New person" : (value.isEmpty ? "Auto-detect patient" : value))
                        .lineLimit(1)
                    Spacer()
                    Image(systemName: "chevron.down")
                }
                .padding(12)
                .frame(maxWidth: .infinity)
                .overlay(RoundedRectangle(cornerRadius: 12).stroke(Color(.separator)))
                .foregroundColor(.primary)
            }
            if manualEntry {
                TextField(tr("New patient name"), text: $value)
                    .textFieldStyle(.roundedBorder)
            }
        }
        .onAppear { members = AppSettings.familyProfiles }
    }
}

File: android-app/app/src/main/java/com/healthdecoder/app/ui/FamilyUi.kt
package com.healthdecoder.app.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.healthdecoder.app.local.LocalRepository
import com.healthdecoder.app.model.FamilyProfile
import kotlinx.coroutines.launch

private val avatarChoices = listOf("👤", "🧑", "👩", "👨", "👵", "👴", "🧒", "👧", "👦", "🧓")
private val sexChoices = listOf("Male", "Female", "Other")

/** Whole years from a YYYY-MM-DD date of birth, or null if unparseable/blank. */
fun familyAge(dob: String): Int? {
    val parts = dob.trim().split("-")
    if (parts.size != 3) return null
    val y = parts[0].toIntOrNull() ?: return null
    val m = parts[1].toIntOrNull() ?: return null
    val d = parts[2].toIntOrNull() ?: return null
    val cal = java.util.Calendar.getInstance()
    val now = java.util.Calendar.getInstance()
    cal.set(y, m - 1, d)
    if (cal.after(now)) return null
    var age = now.get(java.util.Calendar.YEAR) - y
    val beforeBirthday = now.get(java.util.Calendar.MONTH) < (m - 1) ||
        (now.get(java.util.Calendar.MONTH) == (m - 1) && now.get(java.util.Calendar.DAY_OF_MONTH) < d)
    if (beforeBirthday) age--
    return age.takeIf { it in 0..130 }
}

/** One-line "Papa • Father • M • 58y" style summary for a profile. */
fun familySubtitle(p: FamilyProfile): String = buildList {
    if (p.relation.isNotBlank()) add(p.relation)
    if (p.sex.isNotBlank()) add(p.sex.first().uppercase())
    familyAge(p.dateOfBirth)?.let { add("${it}y") }
}.joinToString(" • ")

/**
 * Add or edit one family member. [existing] null = add mode. On save it persists via
 * LocalRepository (a rename cascades to that person's records) and calls [onSaved].
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun FamilyEditDialog(existing: FamilyProfile?, onDismiss: () -> Unit, onSaved: () -> Unit) {
    val context = LocalContext.current
    val scope = rememberCoroutineScope()
    var name by remember { mutableStateOf(existing?.name ?: "") }
    var relation by remember { mutableStateOf(existing?.relation ?: "") }
    var sex by remember { mutableStateOf(existing?.sex ?: "") }
    var dob by remember { mutableStateOf(existing?.dateOfBirth ?: "") }
    var emoji by remember { mutableStateOf(existing?.avatarEmoji ?: "👤") }
    var busy by remember { mutableStateOf(false) }
    var error by remember { mutableStateOf<String?>(null) }

    AlertDialog(
        onDismissRequest = onDismiss,
        title = { Text(if (existing == null) "Add family member" else "Edit ${existing.name}", fontWeight = FontWeight.Bold) },
        text = {
            Column(verticalArrangement = Arrangement.spacedBy(12.dp)) {
                // Avatar picker
                Text(tr("Avatar"), style = MaterialTheme.typography.labelMedium, fontWeight = FontWeight.Bold)
                Row(horizontalArrangement = Arrangement.spacedBy(6.dp), modifier = Modifier.horizontalScroll(rememberScrollState())) {
                    avatarChoices.forEach { e ->
                        val sel = e == emoji
                        Box(
                            modifier = Modifier.size(40.dp).clip(RoundedCornerShape(10.dp))
                                .background(if (sel) MaterialTheme.colorScheme.primaryContainer else MaterialTheme.colorScheme.surfaceVariant.copy(alpha = 0.4f))
                                .border(if (sel) 2.dp else 0.dp, MaterialTheme.colorScheme.primary, RoundedCornerShape(10.dp))
                                .clickable { emoji = e },
                            contentAlignment = Alignment.Center
                        ) { Text(e, fontSize = 20.sp) }
                    }
                }

                OutlinedTextField(
                    value = name, onValueChange = { name = it },
                    label = { Text(tr("Name")) }, singleLine = true, modifier = Modifier.fillMaxWidth()
                )
                OutlinedTextField(
                    value = relation, onValueChange = { relation = it },
                    label = { Text(tr("Relation (e.g. Father, Self)")) }, singleLine = true, modifier = Modifier.fillMaxWidth()
                )

                Text(tr("Sex"), style = MaterialTheme.typography.labelMedium, fontWeight = FontWeight.Bold)
                Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                    sexChoices.forEach { s ->
                        FilterChip(selected = sex == s, onClick = { sex = if (sex == s) "" else s }, label = { Text(s) })
                    }
                }

                OutlinedTextField(
                    value = dob, onValueChange = { dob = it },
                    label = { Text(tr("Birthdate (YYYY-MM-DD)")) },
                    placeholder = { Text("1968-04-15") },
                    singleLine = true,
                    keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Number),
                    supportingText = { familyAge(dob)?.let { Text("Age: $it") } },
                    modifier = Modifier.fillMaxWidth()
                )
                error?.let { Text(it, color = MaterialTheme.colorScheme.error, style = MaterialTheme.typography.bodySmall) }
            }
        },
        confirmButton = {
            Button(
                enabled = !busy && name.isNotBlank(),
                onClick = {
                    scope.launch {
                        busy = true; error = null
                        val ok = if (existing == null) {
                            LocalRepository.addFamilyMember(context, name, relation, sex, dob.trim(), emoji)
                        } else {
                            LocalRepository.updateFamilyMember(context, existing.id, name, relation, sex, dob.trim(), emoji); true
                        }
                        busy = false
                        if (ok) onSaved() else error = "A member with that name already exists."
                    }
                }
            ) { Text(if (existing == null) "Add" else "Save") }
        },
        dismissButton = { TextButton(onClick = onDismiss) { Text(tr("Cancel")) } }
    )
}

/** Lists family members with add/edit/remove. Removing is only offered for people with no records
 *  (rename/merge is the tool for a mis-scanned duplicate). */
@Composable
fun FamilyManagerDialog(onDismiss: () -> Unit, onChanged: () -> Unit) {
    val context = LocalContext.current
    val scope = rememberCoroutineScope()
    var members by remember { mutableStateOf<List<FamilyProfile>>(emptyList()) }
    var editing by remember { mutableStateOf<FamilyProfile?>(null) }
    var adding by remember { mutableStateOf(false) }
    var refresh by remember { mutableStateOf(0) }

    LaunchedEffect(refresh) { members = LocalRepository.familyMembers(context) }

    AlertDialog(
        onDismissRequest = onDismiss,
        title = { Text(tr("Family / Patients"), fontWeight = FontWeight.Bold) },
        text = {
            LazyColumn(modifier = Modifier.fillMaxWidth().heightIn(max = 380.dp), verticalArrangement = Arrangement.spacedBy(6.dp)) {
                items(members, key = { it.id }) { m ->
                    Row(
                        modifier = Modifier.fillMaxWidth().clip(RoundedCornerShape(10.dp))
                            .background(MaterialTheme.colorScheme.surfaceVariant.copy(alpha = 0.4f)).padding(10.dp),
                        verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(10.dp)
                    ) {
                        Text(m.avatarEmoji, fontSize = 22.sp)
                        Column(modifier = Modifier.weight(1f)) {
                            Text(m.name, fontWeight = FontWeight.Bold)
                            familySubtitle(m).takeIf { it.isNotBlank() }?.let {
                                Text(it, style = MaterialTheme.typography.bodySmall, color = MaterialTheme.colorScheme.onSurfaceVariant)
                            }
                        }
                        IconButton(onClick = { editing = m }) { Icon(Icons.Default.Edit, contentDescription = tr("Edit")) }
                        IconButton(onClick = {
                            scope.launch {
                                if (LocalRepository.reportCountFor(context, m.name) > 0) {
                                    // Has records — can't just remove; guide to rename/merge instead.
                                    editing = m
                                } else {
                                    LocalRepository.removeFamilyMember(context, m.id); refresh++; onChanged()
                                }
                            }
                        }) { Icon(Icons.Default.Delete, contentDescription = tr("Remove"), tint = MaterialTheme.colorScheme.error) }
                    }
                }
            }
        },
        confirmButton = {
            Button(onClick = { adding = true }) {
                Icon(Icons.Default.Add, contentDescription = null); Spacer(Modifier.width(6.dp)); Text(tr("Add"))
            }
        },
        dismissButton = { TextButton(onClick = onDismiss) { Text(tr("Done")) } }
    )

    if (adding) FamilyEditDialog(existing = null, onDismiss = { adding = false }, onSaved = { adding = false; refresh++; onChanged() })
    editing?.let { m ->
        FamilyEditDialog(existing = m, onDismiss = { editing = null }, onSaved = { editing = null; refresh++; onChanged() })
    }
}

/**
 * Patient selector for the Scan screen: pick an existing family member or "New person…" (which
 * reveals a name field). [value] is the chosen patient name, empty = auto-detect.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun ScanPatientPicker(value: String, onValueChange: (String) -> Unit, modifier: Modifier = Modifier) {
    val context = LocalContext.current
    var members by remember { mutableStateOf<List<FamilyProfile>>(emptyList()) }
    var menuOpen by remember { mutableStateOf(false) }
    var manualEntry by remember { mutableStateOf(false) }
    LaunchedEffect(Unit) { members = LocalRepository.familyMembers(context) }

    Column(modifier = modifier, verticalArrangement = Arrangement.spacedBy(6.dp)) {
        Box {
            OutlinedButton(onClick = { menuOpen = true }, modifier = Modifier.fillMaxWidth(), shape = RoundedCornerShape(12.dp)) {
                Icon(Icons.Default.Person, contentDescription = null, modifier = Modifier.size(18.dp))
                Spacer(Modifier.width(8.dp))
                Text(
                    when {
                        manualEntry -> "New person"
                        value.isBlank() -> "Auto-detect patient"
                        else -> value
                    },
                    modifier = Modifier.weight(1f)
                )
                Icon(Icons.Default.ArrowDropDown, contentDescription = null)
            }
            DropdownMenu(expanded = menuOpen, onDismissRequest = { menuOpen = false }) {
                DropdownMenuItem(text = { Text(tr("Auto-detect from report")) }, onClick = { manualEntry = false; onValueChange(""); menuOpen = false })
                members.forEach { m ->
                    DropdownMenuItem(
                        text = { Text("${m.avatarEmoji}  ${m.name}" + familySubtitle(m).let { if (it.isNotBlank()) "  ·  $it" else "" }) },
                        onClick = { manualEntry = false; onValueChange(m.name); menuOpen = false }
                    )
                }
                DropdownMenuItem(
                    text = { Text(tr("➕  New person…")) },
                    onClick = { manualEntry = true; onValueChange(""); menuOpen = false }
                )
            }
        }
        if (manualEntry) {
            OutlinedTextField(
                value = value, onValueChange = onValueChange,
                label = { Text(tr("New patient name")) }, singleLine = true,
                modifier = Modifier.fillMaxWidth(), shape = RoundedCornerShape(12.dp)
            )
        }
    }
}
